In [1]:
import pandas as pd
import statsmodels.api as sm
import numpy  as np
from sklearn.preprocessing import StandardScaler

In [2]:
df = pd.read_csv("nb1-insurance.csv")
df.head()

age     sex   bmi  children smoker     region  expenses
0   19  female  27.9         0    yes  southwest  16884.92
1   18    male  33.8         1     no  southeast   1725.55
2   28    male  33.0         3     no  southeast   4449.46
3   33    male  22.7         0     no  northwest  21984.47
4   32    male  28.9         0     no  northwest   3866.86

In [92]:
# Variable objetivo
Y = df["expenses"]
#Variables predictivas
X = df[["age","sex","bmi","children","smoker"]]

In [ ]:
pd.read

In [93]:
columns_names = df.columns.values.tolist()
columns_names

['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'expenses']

In [94]:
dummy_sex = pd.get_dummies(df["sex"], prefix="sex").iloc[:,1:]
dummy_sex.head()

sex_male
0         0
1         1
2         1
3         1
4         1

In [95]:
dummy_smoker = pd.get_dummies(df["smoker"], prefix="smoker").iloc[:,1:]
dummy_smoker.head()

smoker_yes
0           1
1           0
2           0
3           0
4           0

In [96]:
columns_names = df.columns.values.tolist()
df_new = df[columns_names].join(dummy_sex)
columns_names = df_new.columns.values.tolist()
df_new = df_new[columns_names].join(dummy_smoker)
df_new.head()

age     sex   bmi  children smoker     region  expenses  sex_male  \
0   19  female  27.9         0    yes  southwest  16884.92         0   
1   18    male  33.8         1     no  southeast   1725.55         1   
2   28    male  33.0         3     no  southeast   4449.46         1   
3   33    male  22.7         0     no  northwest  21984.47         1   
4   32    male  28.9         0     no  northwest   3866.86         1   

   smoker_yes  
0           1  
1           0  
2           0  
3           0  
4           0

In [97]:
feature_cols = ["age","bmi","children","sex_male","smoker_yes"]
# Variable objetivo
Y = df_new["expenses"]
#Variables predictivas
X = df_new[feature_cols]

In [98]:
# Normalizando variables
scale = StandardScaler()
X_scaled = scale.fit_transform(X[feature_cols].as_matrix())


In [99]:
# Estimacion variables sin normalizar
est = sm.OLS(Y, X).fit()
est.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               expenses   R-squared:                       0.873
Model:                            OLS   Adj. R-squared:                  0.872
Method:                 Least Squares   F-statistic:                     1826.
Date:                Tue, 11 Sep 2018   Prob (F-statistic):               0.00
Time:                        11:46:59   Log-Likelihood:                -13627.
No. Observations:                1338   AIC:                         2.726e+04
Df Residuals:                    1333   BIC:                         2.729e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
age          198.3998     11.581     17.131      0.000     175.680     221.120
bmi           37.5784     16.611      2.262      0.024       4.992      70.165
children     250.8414    144.670      1.734      0.083     -32.964     534.647
sex_male    -707.6684    349.415     -2.025      0.043   -1393.132     -22.205
smoker_yes  2.339e+04    434.988     53.763      0.000    2.25e+04    2.42e+04
==============================================================================
Omnibus:                      275.545   Durbin-Watson:                   2.069
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              631.013
Skew:                           1.133   Prob(JB):                    9.49e-138
Kurtosis:                       5.487   Cond. No.                         128.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [100]:
# Estimacion variables normalizadas
est2 = sm.OLS(Y,X_scaled).fit()
est2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               expenses   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.338
Method:                 Least Squares   F-statistic:                     137.7
Date:                Tue, 11 Sep 2018   Prob (F-statistic):          8.13e-118
Time:                        11:46:59   Log-Likelihood:                -14727.
No. Observations:                1338   AIC:                         2.946e+04
Df Residuals:                    1333   BIC:                         2.949e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1          3619.5908    402.544      8.992      0.000    2829.903    4409.279
x2          1965.6982    402.434      4.885      0.000    1176.225    2755.171
x3           571.9154    399.976      1.430      0.153    -212.736    1356.567
x4           -64.3371    401.299     -0.160      0.873    -851.583     722.909
x5          9613.3311    400.817     23.984      0.000    8827.030    1.04e+04
==============================================================================
Omnibus:                      301.233   Durbin-Watson:                   0.360
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              720.877
Skew:                           1.215   Prob(JB):                    2.91e-157
Kurtosis:                       5.651   Cond. No.                         1.14
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [101]:
print("Variables sin normalizar:\n",est.params)
print()
print("Varialbes normalizadas\n",est2.params)

Variables sin normalizar:
 age             198.399764
bmi              37.578382
children        250.841441
sex_male       -707.668406
smoker_yes    23386.227638
dtype: float64

Varialbes normalizadas
 x1    3619.590806
x2    1965.698187
x3     571.915447
x4     -64.337115
x5    9613.331140
dtype: float64


In [102]:
print("R^2 variable sin normalizar: ",est.rsquared)
print("R^2 varialbes normalizadas: ",est2.rsquared)

R^2 variable sin normalizar:  0.8726278733269306
R^2 varialbes normalizadas:  0.34052292290708386


In [103]:
def predict_expenses(age, bmi, children,sex_male,smoker_yes):
    return 198.399764*age  +  37.578382*bmi + 250.841441*children - 707.668406*sex_male + 23386.227638*smoker_yes

In [104]:
def predict_expenses_normalize(age, bmi, children,sex_male,smoker_yes):
    return 3619.590806*age  +  1965.698187*bmi + 571.915447*children - 64.337115*sex_male + 9613.331140*smoker_yes

In [105]:
df_new["prediction"] = predict_expenses(df_new["age"],df_new["bmi"],df_new["children"],df_new["sex_male"],df_new["smoker_yes"])
df_new["prediction_normalize"] = predict_expenses_normalize(df_new["age"],df_new["bmi"],df_new["children"],df_new["sex_male"],df_new["smoker_yes"])

In [106]:
df_new.head()

age     sex   bmi  children smoker     region  expenses  sex_male  \
0   19  female  27.9         0    yes  southwest  16884.92         0   
1   18    male  33.8         1     no  southeast   1725.55         1   
2   28    male  33.0         3     no  southeast   4449.46         1   
3   33    male  22.7         0     no  northwest  21984.47         1   
4   32    male  28.9         0     no  northwest   3866.86         1   

   smoker_yes    prediction  prediction_normalize  
0           1  28204.260012         133228.535871  
1           0   4384.518099         132100.811561  
2           0   6840.135915         167867.991965  
3           0   6692.553077         164003.508328  
4           0   6727.139282         172571.246281

In [107]:
print("Gastos de salud: $",predict_expenses(55,42.6,0,1,1))
print("Gastos de salud, var. normalizada: $",predict_expenses_normalize(55,42.6,0,1,1))

Gastos de salud: $ 35191.3853252
Gastos de salud, var. normalizada: $ 292365.2311212


* Desde un punto de vista del emprendimiento, ¿qué podría ser el valor añadido de este tipo de modelo predictivo? a quién podría servir?
    * El valor añadido para este modelo predictivo podria ser, un modelo que calcula posibles gastos para el individuo, le serviria a personas que quieran manejar o mejorar sus finanzas o personas que quieran saber cuanto mejoraria si fuma o no, o si podria solventar el tener otro hijo, etc

----------------------
* Desde un punto de vista científico y ético, ¿qué críticas podría emitir?
    * Desde el punto de vista científico podria decir que este metodo tiene un R^2 de 0.87 esto quiere decir que la prediccion es un 87% acertada. Desde el punto de vista ético podria decir que la estimacion realizada es buena, podria servir como un punto de referencia para la persona que quiera saber los gastos que tiene, pero no es completamente acertada